Add Tasks doc (#1044)
lit/lit.dev: packages/lit-dev-content/site/docs/v3/data/task.md

While Lit's reactive update lifecycle is batched and asynchronous, Lit templates always render _synchronously_. The data used in a template must be readable at the time of rendering. To render async data in a Lit component, you must wait for the data to be ready, store it so that it's readable, then trigger a new render which can use the data synchronously. Considerations must often be made on what to render while the data is being fetched, or when the data fetch fails as well.

The `@lit/task` package provides a `Task` reactive controller to help manage this async data workflow.

`Task` is a controller that takes an async task function and runs it either manually or automatically when its arguments change. Task stores the result of the task function and updates the host element when the task function completes so the result can be used in rendering.

### Example

This is an example of using `Task` to call an HTTP API via [`fetch()`](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API). The API is called whenever the `productId` parameter changes, and the component renders a loading message when the data is being fetched.

{% switchable-sample %}

```ts

> import {Task} from '@lit-labs/task';

Reviewer: Nit: `@lit-labs/task` -> `@lit/task`
(and for other code samples)

```suggestion
import {Task} from '@lit/task';
```

class MyElement extends LitElement {
  @property() productId?: string;

  private _productTask = new Task(this, {
    task: async ([productId], {signal}) => {

>       const response = await fetch(`http://example.com/product/${productId}`, {signal});

Reviewer: Would it be useful if examples worked in the lit.dev playground without too many changes?

I wanted to try this out, but it required me fixing the URL to `https:` and adding `mode: "no-cors",` to the fetch options. Then the URL always seems to return 404 (which I think makes sense because it's an example URL).

Async data is data that's not available immediately, but may be available at some time in the future. For example, instead of a value like a string or an object that's usable synchronously, a Promise provides a value in the future.

Async data is usually returned from an async API, which can come in a few forms:
- Promises and async functions which return Promises
- Functions that accept callbacks
- Objects that emit events
- Various async patterns libraries like observables and signals

## What is a task?

At the core of the Task controller is the concept of a "task" itself.

A task is an async operation which does some work to produce data. A task can be in a few different states (initial, pending, complete, and error) and can take parameters.

A task is a generic concept and could represent any async operation. They apply best when there is a request/response structure, such as a network fetch, database query, or waiting for a single event in response to some action. They're less applicable to spontaneous or streaming operations like an open-ended stream of events, a streaming database response, etc.

## Installation

```bash
npm i @lit/task
```

## Usage

> `Task` is a [reactive controller](../../composition/controllers/), so it can respond to and trigger updates to Lit's reactive update lifecycle.

Reviewer: I think we prefer server relative URLs. (That's the convention. I forget the precise reason).
```suggestion
`Task` is a [reactive controller](/docs/v3/composition/controllers/), so it can respond to and trigger updates to Lit's reactive update lifecycle.
```

You'll generally have one Task object for each logical task that your component needs to perform. Install tasks as fields on your class:

{% switchable-sample %}

```ts
class MyElement extends LitElement {
  private _myTask = new Task(this, {/*...*/});
}
```

```js
class MyElement extends LitElement {
  _myTask = new Task(this, {/*...*/});
}
```

{% endswitchable-sample %}

As a class field, the task status and value are easily available:

```ts

> this._task.value

Reviewer: Nit: in sentence above introduce `status`, and `value`. Could this example also have `this._task.status`?

```

### The task function

The most critical part of a task declaration is the _task function_. This is the function that does the actual work.

The task function is given in the `task` option. The Task controller will automatically call the task function with arguments, which are supplied with a separate `args` callback. Arguments are checked for changes and the task function is only called if the arguments have changed.

The task function takes the task arguments as an _array_ passed as the first parameter, and an options argument as the second parameter:

```ts
new Task(this, {
  task: async ([arg1, arg2], {signal}) => {
    // do async work here
  },
  args: () => [this.field1, this.field2]
})
```

The task function's args array and the args callback should be the same length.

{% aside "positive" "no-header" %}

> Write the `task` and `args` functions as arrow function so that the `this` reference points to the host element.

Reviewer: ```suggestion
Write the `task` and `args` functions as arrow functions so that the `this` reference points to the host element.
```

    `;
  }

  _onClick() {
    this._getDataTask.run();
  }
}
```

{% endswitchable-sample %}

In manual mode you can provide new arguments directly to `run()`:

```ts
this._task.run('arg1', 'arg2');
```

If arguments are not provided to `run()`, they are gathered from the `args` callback.

### Aborting tasks

The second argument to the task function is an options object that carries an [`AbortSignal`](https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal) in the `signal` property. This can be used to cancel a pending task because a new task run has started.

The easiest way to use a signal is to forward it to an API that accepts it, like `fetch()`.

{% switchable-sample %}

```ts
  private _task = new Task(this, {
    task: async (args, {signal}) => {
      const response = await fetch(someUrl, {signal});
      // ...
    },
  });
```

```js
  _task = new Task(this, {
    task: async (args, {signal}) => {
      const response = await fetch(someUrl, {signal});
      // ...
    },
  });
```

{% endswitchable-sample %}

Forwarding the signal to `fetch()` will cause the browser to cancel the network request if the signal is aborted.

> You can also check if a signal has been aborted in your task function. It's often useful to race a Promise against a signal:

Reviewer: This says "race a signal against a promise", but then the example doesn't look like it's racing. I'd expect some kind of `Promise.race` usage.

Author: Promise.race is a way to race multiple promises against each other, and it's not the only way to race promises. In this case we're not racing multiple promises though, but an abort signal and promises.

Author: fixed

{% switchable-sample %}

```ts
  private _task = new Task(this, {
    task: async ([arg1], {signal}) => {
      const firstResult = await doSomeWork(arg1);
      signal.throwIfAborted();
      const secondResult = await doMoreWork(firstResult);
      signal.throwIfAborted();
      return secondResult;
    },
  });
```

```js
  _task = new Task(this, {
    task: async ([arg1], {signal}) => {
      const firstResult = await doSomeWork(arg1);
      signal.throwIfAborted();
      const secondResult = await doMoreWork(firstResult);
      signal.throwIfAborted();
      return secondResult;
    },
  });
```

{% endswitchable-sample %}
### Task chaining

> Sometimes you want to run one task when another task completes. To do this you can use the value of a task as an argument to the other:

Reviewer: ```suggestion
Sometimes you want to run one task when another task completes. To do this you can use the value of a task as an argument to another task:
```

}
```

```js
class MyElement extends LitElement {
  _getDataTask = new Task(this, {
    task: ([dataId]) => getData(dataId),
    args: () => [this.dataId],
  });

  _processDataTask = new Task(this, {
    task: ([data, param]) => processData(data, param),
    args: () => [this._getDataTask.value, this.param],
  });
}
```

{% endswitchable-sample %}

This can be useful if the tasks have different arguments so that the chained task may run without the first task running again. In this case it'll use the first task like a cache.

You can also often use one task function and await intermediate results:

{% switchable-sample %}

```ts
class MyElement extends LitElement {
  private _getDataTask = new Task(this, {
    task: ([dataId, param]) => {
      const data = await getData(dataId);
      return processData(data, param);
    },
    args: () => [this.dataId, this.param],
  });
}
```

```js
class MyElement extends LitElement {
  _getDataTask = new Task(this, {
    task: ([dataId, param]) => {
      const data = await getData(dataId);
      return processData(data, param);
    },
    args: () => [this.dataId, this.param],
  });
}
```

{% endswitchable-sample %}

### Race conditions

> A task function can be called while previous task calls are still pending. In this case the `AbortSignal` passed to previous runs will be aborted and the `Task` controller will ignore the result of the previous task runs and only handle the most recent run.

Reviewer: Maybe this section can be removed? It seems like a duplicate of the `### Aborting tasks` section earlier in the documentation page.

- Runs task functions when arguments change
- Tracks the task status (initial, pending, complete, or error)
- Saves the last completion value or error of the task function
- Triggers a host update when the task changes status
- Handles race conditions, ensuring that only the latest task invocation completes the task
- Renders the correct template for the current task status
- Allows aborting tasks with an [`AbortController`](https://developer.mozilla.org/en-US/docs/Web/API/AbortController)

This removes most of the boilerplate for correctly using async data from your code, and ensures robust handling of race conditions and other edge-cases.

## What is async data?

Async data is data that's not available immediately, but may be available at some time in the future. For example, instead of a value like a string or an object that's usable synchronously, a promise provides a value in the future.

Async data is usually returned from an async API, which can come in a few forms:
- Promises or async functions, like `fetch()`
- Functions that accept callbacks
- Objects that emit events, such as DOM events
- Libraries like observables and signals

> The Task controller deals in promises, so no matter the shape of your async API you'll adapt it to promises to use with Task.

Reviewer 2: ```suggestion
The Task controller deals in promises, so no matter the shape of your async API you can adapt it to promises to use with Task.
```

  _getDataTask = new Task(
    this,
    {
      task: async () =>
        const response = await fetch(`example.com/data/`);
        return response.json();
      ),
      args: () => []
    }
  );

  render() {
    return html`
      <button @click=${this._onClick}>Get Data</button>
    `;
  }

  _onClick() {
    this._getDataTask.run();
  }
}
```

{% endswitchable-sample %}

In manual mode you can provide new arguments directly to `run()`:

```ts
this._task.run('arg1', 'arg2');
```

If arguments are not provided to `run()`, they are gathered from the `args` callback.

### Aborting tasks

A task function can be called while previous task runs are still pending. In these cases the result of the pending task runs will be ignored, and you should try to cancel any outstanding work or network I/O in order to save resources.

You can do with with the [`AbortSignal`](https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal) that is passed in the `signal` property of the second argument to the task function. When a pending task run is superseded by a new run, the `AbortSignal` that was passed to the pending run is aborted to signal the task run to cancel any pending work.

> `AbortSignal` doesn't cancel any work automatically - it is just a signal. To cancel some work you must either do it yourself by checking the signal, or forward the signal to another API that accepts `AbortSignal`s like [`fetch()`](https://developer.mozilla.org/en-US/docs/Web/API/fetch) or [`addEventListner()`](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener).

Reviewer 2: ```suggestion
`AbortSignal` doesn't cancel any work automatically - it is just a signal. To cancel some work you must either do it yourself by checking the signal, or forward the signal to another API that accepts `AbortSignal`s like [`fetch()`](https://developer.mozilla.org/en-US/docs/Web/API/fetch) or [`addEventListener()`](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener).
```
